File: python/mirage/core/lancedb/stat.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.lancedb import LanceDBAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.lancedb.query import table_exists
from mirage.core.lancedb.read import read
from mirage.core.lancedb.scope import ScopeLevel, detect_scope
from mirage.types import FileStat, FileType, PathSpec

_IMAGE_TYPES = {
    "png": FileType.IMAGE_PNG,
    "jpg": FileType.IMAGE_JPEG,
    "jpeg": FileType.IMAGE_JPEG,
    "gif": FileType.IMAGE_GIF,
}


def _name_of(path: PathSpec) -> str:
    stripped = path.original.rstrip("/")
    return stripped.rsplit("/", 1)[-1] or "/"


async def stat(
    accessor: LanceDBAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> FileStat:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    config = accessor.config
    scope = detect_scope(path, config)

    if scope.level == ScopeLevel.UNKNOWN:
        raise FileNotFoundError(path.original)

    if scope.table and not await table_exists(accessor, scope.table):
        raise FileNotFoundError(path.original)

    if scope.level in (ScopeLevel.ROOT, ScopeLevel.GROUP_DIR,
                       ScopeLevel.SEARCH_DIR, ScopeLevel.SEARCH_RESULTS):
        return FileStat(name=_name_of(path), type=FileType.DIRECTORY)

    data = await read(accessor, path, index)
    if scope.blob:
        file_type = _IMAGE_TYPES.get(config.blob_ext, FileType.BINARY)
    else:
        file_type = FileType.TEXT
    return FileStat(name=_name_of(path), size=len(data), type=file_type)

File: typescript/packages/core/src/core/lancedb/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LanceDBAccessor } from '../../accessor/lancedb.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { FileStat, FileType, PathSpec } from '../../types.ts'
import { rstripSlash } from '../../utils/slash.ts'
import { read } from './read.ts'
import { ScopeLevel, detectScope } from './scope.ts'

function notFound(p: string): Error {
  const err = new Error(p) as Error & { code?: string }
  err.code = 'ENOENT'
  return err
}

const IMAGE_TYPES: Record<string, FileType> = {
  png: FileType.IMAGE_PNG,
  jpg: FileType.IMAGE_JPEG,
  jpeg: FileType.IMAGE_JPEG,
  gif: FileType.IMAGE_GIF,
}

function nameOf(spec: PathSpec): string {
  const stripped = rstripSlash(spec.original)
  const last = stripped.split('/').pop()
  return last === undefined || last === '' ? '/' : last
}

export async function stat(
  accessor: LanceDBAccessor,
  path: PathSpec | string,
  index?: IndexCacheStore,
): Promise<FileStat> {
  const spec = typeof path === 'string' ? PathSpec.fromStrPath(path) : path
  const config = accessor.config
  const scope = detectScope(spec, config)

  if (scope.level === ScopeLevel.UNKNOWN) throw notFound(spec.original)

  if (scope.table !== null) {
    const tables = await accessor.driver.listTables()
    if (!tables.includes(scope.table)) throw notFound(spec.original)
  }

  if (scope.level === ScopeLevel.ROOT || scope.level === ScopeLevel.GROUP_DIR) {
    return new FileStat({ name: nameOf(spec), type: FileType.DIRECTORY })
  }

  const data = await read(accessor, spec, index)
  const fileType = scope.blob ? (IMAGE_TYPES[config.blobExt] ?? FileType.BINARY) : FileType.TEXT
  return new FileStat({ name: nameOf(spec), size: data.length, type: fileType })
}
